src/main/java/com/celements/web/contextmenu/ContextMenuItem.java

```diff
@@ -21,77 +21,49 @@
 
 import java.math.BigDecimal;
 import java.util.Arrays;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 
 import org.apache.velocity.VelocityContext;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
-import org.xwiki.context.Execution;
+import org.xwiki.velocity.XWikiVelocityException;
 
 import com.celements.sajson.Builder;
+import com.celements.velocity.VelocityContextModifier;
+import com.celements.velocity.VelocityService;
+import com.google.common.base.Strings;
 import com.xpn.xwiki.XWikiContext;
 import com.xpn.xwiki.objects.BaseObject;
 import com.xpn.xwiki.web.Utils;
 
 public class ContextMenuItem {
 
-  public static final String CONTEXTMENUITEM_CLASSNAME = "Celements2.ContextMenuItemClass";
-
   private static final Logger LOGGER = LoggerFactory.getLogger(ContextMenuItem.class);
 
-  private String cmiLink;
-  private String cmiText;
-  private String cmiIcon;
-  private String shortcut;
-
-  private String elemId;
-  private String origElemId;
+  private final BaseObject menuItemObj;
```

Comment:
keep POJO setup. Move towards imputable objects to allow caching later on

Reply:
not sure I understand what you mean by "keep pojo setup". cloned obj though to ensure effective immutability.

Reply:
The target software architecture is a XWiki independent Context-Menu. Thus XWiki-classes should only be used to get read by the builder to fill in the ContextMenuItem values. Nevertheless We want to be able to create ContextMenuItem directly out ouf java without backing them by BaseObjects. Moreover the Java generated ContextMenuItem should only be generated as immutable objects and then kept in Memory. Thus the caching of the evaluated velocity scripts must be in different objects from different classes (mybe RenderedContextMenuItem). The RenderedContextMenuItem is only cachable in a user/request context. The ContextMenuItem will be cachable system-wide.

Reply:
Thanks for clarification. I decoupled the `ContextMenuItem` from the `BaseObject` and introduced a constructor without it to also allow CMI creation without storage instances, according to your vision.

I agree that the current CMI is not suitable for system wide caching, I noted this in it's javadoc. At the time we're going to add a system wide CMI-cache we'll have to reconsider it's impact thoroughly (I remember the headache with the document cache) and potentially refactor the CMI to our needs. I feel venturing into this future feature is out of scope at this point.